### PR TITLE
Fallback to 5.1 Wide Layout on old drivers

### LINKED ISSUE #22584: Possible Regression with EAC3 Audio Decoding in v20
### Describe the bug
Here is a clear and concise description of what the problem is:

I have an older Yamaha AVR with only standard 5.1 Dolby AC3 and DTS (both 5.1) decoding.  No TrueHD or 7.1 decoding.  

I have Kodi and Windows setup with 5.1 audio and Kodi setup to bitstream standard 5.1 AC3 and DTS (using DTS Core).  Everything else gets decoded (and downmixed as necessary) by Kodi and output as PCM 5.1 (or 2.0/1 if it's a stereo track) to the receiver via WASPI over HDMI.

In v19.5, EAC3 5.1 was working fine as expected, it would be decoded and output as 5.1 PCM to the receiver, channels properly mapped.

In v20.0, this no longer works. The same EAC3 5.1 is now output as 2 channel PCM.  Regular 5.1 AC3 and DTS core still properly bitstream as expected to the AVR, as they did before in 19.5.  I also have some videos with WMA5.1 audio to test, and those also still get properly decoded and output as 5.1 PCM as well, as before.  I also have some 7.1 TrueHD files for testing, and those also decode and remap and output to the AVR to 5.1 PCM fine too.  It's just the EAC3 files that don't work properly anymore.  It does that with the released v20.0, and also with the latest nightly build.

It was working fine as expected in 19.5 and decoding and outputting these audio steams as 5.1 PCM, but after installing v20.0 over top of my previous installation, I noticed that the audio was only being output as 2.0 and not 5.1.  I confirmed that it still works fine in 19.5 by simply reinstalling that version over top of the v20, and launching the same video file. It is output as 5.1 PCM as expected.  Then simply reinstalling v20 overtop once again, and the same audio is  now output as PCM 2.0 again.  Nothing else changed, other than installing one version or the other overtop of the existing installation.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I expected that videos with 5.1 EAC3 audio would be decoded and output to the AVR over HDMI by Kodi as 5.1 PCM.

## Actual Behavior

Audio is output to the AVR as PCM 2.0

## Possible Fix

I don't know, it works fine in 19.5 and not in 20.0

### To Reproduce

1. Launch a file with EAC3 5.1 audio in v19.5 and watch the channel lights on the AVR.  All 5 channel lights illuminate
2. Install v20.0 over top, launch it immediately afterwards, launch the same video file, watch the AVR, only 2 lights (stereo illuminate)
3. Reinstall 19.5 over top again, launch it, test the same file, it is now decoding as 5.1 PCM to the AVR again as expected.
4. Reinstall v20.0 overtop, repeat playback and it now only outputs PCM2.0
5. This indicates there has to be some kind of regression or bug in 20.0 as it is not decoding the EAC3 file properly as it did before in v19

### Debuglog

The debuglog can be found here:

Debug log for v19.5:
https://hastebin.com/utiqirudul.yaml

Debug log for 20.0 playing the same file:
https://hastebin.com/ukicavuluq.yaml

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [x ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Win 10 x64
 - Kodi version: 20.0
 - Also does the same thing in latest nightly build (20230123)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by fritsch**
Should mitigate: #22584

### CODE REVIEW

**Comment by fritsch** [comment]
Jenkins win64 builder did not like me, let's see if this version can be built.

File: xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp [hunk 1]
```diff
@@ -771,7 +771,26 @@ bool CAESinkWASAPI::InitializeExclusive(AEAudioFormat &format)
...
+      {
+        bool skip = true;
+        if (requestedChannels == 6 && wfxex.Format.nChannels == 6)
+        {
+          // For old sound cards it might look as follows
+          // Works: Channel Layout: FL, FR, FC, LFE, BL, BR
+          // Does not work: FL, FR, FC, LFE, SL, SR
+          // in the mask comparison SL and SR remain as driver only
+          // returns BL and BR channel
+          CLog::Log(LOGINFO, "Res: {} SL|SR {} BL|BR", res, (int) (SPEAKER_SIDE_LEFT | SPEAKER_SIDE_RIGHT) , (int) (SPEAKER_BACK_LEFT | SPEAKER_BACK_RIGHT));
```
Comment (thexai) [inline xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp:783]:
```suggestion
          CLog::LogF(LOGINFO, "Res: {} SL|SR {} BL|BR", res, (int) (SPEAKER_SIDE_LEFT | SPEAKER_SIDE_RIGHT) , (int) (SPEAKER_BACK_LEFT | SPEAKER_BACK_RIGHT));
```

File: xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp [hunk 2]
```diff
@@ -771,7 +771,26 @@ bool CAESinkWASAPI::InitializeExclusive(AEAudioFormat &format)
...
+          if (res == (SPEAKER_SIDE_LEFT | SPEAKER_SIDE_RIGHT) &&
+              (wfxex.dwChannelMask & (SPEAKER_BACK_LEFT | SPEAKER_BACK_RIGHT) != 0))
+          {
+            CLog::Log(LOGDEBUG, "Falling back to Layout with SL SR swaped with BL BR");
```
Comment (thexai) [inline xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp:787]:
```suggestion
            CLog::LogF(LOGDEBUG, "Falling back to Layout with SL SR swaped with BL BR");
```

Comment (fritsch) [reply xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp]: Thanks done.

File: xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp (hunk 1 above)
Comment (fritsch) [reply xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp]: Thanks done.

File: xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp [hunk 3]
```diff
@@ -771,7 +771,28 @@ bool CAESinkWASAPI::InitializeExclusive(AEAudioFormat &format)
...
+        if (requestedChannels == 6 && wfxex.Format.nChannels == 6)
+        {
+          // For old sound cards it might look as follows
+          // Works: Channel Layout: FL, FR, FC, LFE, BL, BR
+          // Does not work: FL, FR, FC, LFE, SL, SR
+          // in the mask comparison SL and SR remain as driver only
+          // returns BL and BR channel
+          CLog::LogF(LOGDEBUG, "Fallback Matching Res: {} SL|SR {} BL|BR", res,
+                     (int)(SPEAKER_SIDE_LEFT | SPEAKER_SIDE_RIGHT),
+                     (int)(SPEAKER_BACK_LEFT | SPEAKER_BACK_RIGHT));
```
Comment (thexai) [inline xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp:785]: Seems this log line has 3 parameters and only 2 placeholders? 

`"Fallback Matching Res: {} SL|SR {} BL|BR {}"`  ??

Also cast to int not seems necessary as SPEAKER_SIDE_LEFT / RIGHT are also ints

Comment (fritsch) [reply xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp:785]: Yes, sorry for that. Blind coding without runtime testing ... is not the best thing to keep jenkins resources low ;-). Thanks.

**Comment by CrystalP** [comment]
Jenkins please build this

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by fritsch** [comment]
While the code is correct (TM), it does not work.

**Comment by thexai** [comment]
If with latest change not works, maybe this is more generic approach:

```diff
diff --git a/xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp b/xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp
index 1f35ce3c18..e808bac6e5 100644
--- a/xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp
+++ b/xbmc/cores/AudioEngine/Sinks/AESinkWASAPI.cpp
@@ -730,6 +730,7 @@ bool CAESinkWASAPI::InitializeExclusive(AEAudioFormat &format)
   unsigned int requestedChannels = 0;
   unsigned int noOfCh = 0;
   uint64_t desired_map = 0;
+  bool matchNoChannelsOnly = false;
 
   if (SUCCEEDED(hr))
   {
@@ -762,16 +763,34 @@ bool CAESinkWASAPI::InitializeExclusive(AEAudioFormat &format)
     // as the last resort try stereo
     if (layout == ARRAYSIZE(layoutsList))
     {
-      wfxex.dwChannelMask = SPEAKER_FRONT_LEFT | SPEAKER_FRONT_RIGHT;
-      wfxex.Format.nChannels = 2;
+      if (matchNoChannelsOnly)
+      {
+        wfxex.dwChannelMask = SPEAKER_FRONT_LEFT | SPEAKER_FRONT_RIGHT;
+        wfxex.Format.nChannels = 2;
+      }
+      else
+      {
+        matchNoChannelsOnly = true;
+        layout = -1;
+        CLog::Log(LOGWARNING, "AESinkWASAPI: Match only number of audio cahnnels as fallback");
+        continue;
+      }
     }
     else if (layout >= 0)
     {
       wfxex.dwChannelMask = CAESinkFactoryWin::ChLayoutToChMask(layoutsList[layout], &noOfCh);
       wfxex.Format.nChannels = noOfCh;
       int res = desired_map & wfxex.dwChannelMask;
-      if (res != desired_map)
-        continue; // output channel layout doesn't match input channels
+      if (matchNoChannelsOnly)
+      {
+        if (noOfCh != requestedChannels)
+          continue; // number of channels doesn't match requested channels
+      }
+      else
+      {
+        if (res != desired_map)
+          continue; // output channel layout doesn't match input channels
+      }
     }
     CAEChannelInfo foundChannels;
     CAESinkFactoryWin::AEChannelsFromSpeakerMask(foundChannels, wfxex.dwChannelMask);

```

Test all layout list as before but instead of fallback to stereo test again all list but matching number of channels only (and supported on system).

**Comment by fritsch** [comment]
Feel free to superseed. You are much nearer towards the Windows platform than I am.

**Comment by fritsch** [comment]
Other than that - the current one worked. Let's do what you prefer @thexai

**Comment by thexai** [comment]
The code that worked is this https://github.com/thexai/xbmc/commit/76e5236751306a4d144766627404e135af8eedd8

**Comment by fritsch** [comment]
Closing as superseeded.

**Pull request closed without merging**